### DISCUSSION TITLE
"python3 -m pip install pi-heif" fails with "error: implicit declaration of function ‘heif_image_handle_get_preferred_decoding_colorspace’" when installing picframe

### DISCUSSION THREAD

**Discussion by oggreeb** [bigcat88/pillow_heif#280] [Q&A] [answered]
Hello. I'm using helgeerbe's picframe package to install a digital picture frame onto a Raspberry Pi Zero 2W using RPi Bookworm 32-bit Lite. I'm getting this error when attempting to install. It's consistant across multiple attempts and with  "python3 -m pip install --upgrade pillow-heif --no-binary :all:"   --

------------------

(venv_picframe) pi@rpiz2w:~ $ python3 -m pip install -U pip
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
Requirement already satisfied: pip in ./venv_picframe/lib/python3.11/site-packages (24.2)
(venv_picframe) pi@rpiz2w:~ $ python3 -m pip install pi-heif
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
Collecting pi-heif
  Using cached pi_heif-0.18.0.tar.gz (16.2 MB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Requirement already satisfied: pillow>=10.1.0 in ./venv_picframe/lib/python3.11/site-packages (from pi-heif) (10.4.0)
Building wheels for collected packages: pi-heif
  Building wheel for pi-heif (pyproject.toml) ... error
  error: subprocess-exited-with-error

  × Building wheel for pi-heif (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [33 lines of output]
      running bdist_wheel
      running build
      running build_py
      creating build
      creating build/lib.linux-armv7l-cpython-311
      creating build/lib.linux-armv7l-cpython-311/pi_heif
      copying pi_heif/misc.py -> build/lib.linux-armv7l-cpython-311/pi_heif
      copying pi_heif/options.py -> build/lib.linux-armv7l-cpython-311/pi_heif
      copying pi_heif/__init__.py -> build/lib.linux-armv7l-cpython-311/pi_heif
      copying pi_heif/HeifImagePlugin.py -> build/lib.linux-armv7l-cpython-311/pi_heif
      copying pi_heif/constants.py -> build/lib.linux-armv7l-cpython-311/pi_heif
      copying pi_heif/heif.py -> build/lib.linux-armv7l-cpython-311/pi_heif
      copying pi_heif/as_plugin.py -> build/lib.linux-armv7l-cpython-311/pi_heif
      copying pi_heif/_deffered_error.py -> build/lib.linux-armv7l-cpython-311/pi_heif
      copying pi_heif/AvifImagePlugin.py -> build/lib.linux-armv7l-cpython-311/pi_heif
      copying pi_heif/_version.py -> build/lib.linux-armv7l-cpython-311/pi_heif
      copying pi_heif/_lib_info.py -> build/lib.linux-armv7l-cpython-311/pi_heif
      running build_ext
      Looking for `libheif` using pkg-config.
      Checking for include file 'heif.h' in '/usr/include/arm-linux-gnueabihf'
      Checking for include file 'heif.h' in '/usr/include/arm-linux-gnueabihf/libheif'
      Found 'heif.h' in '/usr/include/arm-linux-gnueabihf/libheif'
      building '_pi_heif' extension
      creating build/temp.linux-armv7l-cpython-311
      creating build/temp.linux-armv7l-cpython-311/pi_heif
      arm-linux-gnueabihf-gcc -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fwrapv -O2 -fPIC -I/usr/include/arm-linux-gnueabihf -I/usr/include -I/home/pi/venv_picframe/include -I/usr/local/include -I/home/pi/venv_picframe/include -I/usr/include/python3.11 -c pi_heif/_pi_heif.c -o build/temp.linux-armv7l-cpython-311/pi_heif/_pi_heif.o -Ofast -Werror
      pi_heif/_pi_heif.c: In function ‘_load_file’:
      pi_heif/_pi_heif.c:1366:21: error: implicit declaration of function ‘heif_image_handle_get_preferred_decoding_colorspace’ [-Werror=implicit-function-declaration]
       1366 |             error = heif_image_handle_get_preferred_decoding_colorspace(handle, &colorspace, &chroma);
            |                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      pi_heif/_pi_heif.c:1366:21: error: incompatible types when assigning to type ‘struct heif_error’ from type ‘int’
      cc1: all warnings being treated as errors
      error: command '/usr/bin/arm-linux-gnueabihf-gcc' failed with exit code 1
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for pi-heif
Failed to build pi-heif
ERROR: ERROR: Failed to build installable wheels for some pyproject.toml based projects (pi-heif)
(venv_picframe) pi@rpiz2w:~ $

----------------------

Thank you in advance for any advice.

**Comment by bigcat88** [comment, marked as answer]
Hello. What version of `libheif` do you have installed?

The minimum required version of `libehif` is `1.17.0`

**Reply by oggreeb** [reply]
That was the problem -- the RPi Bookworm latest packages only had 1.15.1. I also installed png, jpeg, tiff packages that didn't come automatically with the lite distro. I then grabbed 1.18.2 from the GH code base and it compiled fine. Thank you!
